crates/viewer/re_time_panel/src/data_density_graph.rs

```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+    if let Some(component_name) = component_name {
+        let chunk = Arc::new(chunk.range(&query, component_name));
+        let Some(num_events) = chunk.num_events_for_component(component_name) else {
+            return;
+        };
+        let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+            return;
+        };
+        visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+    } else {
+        for component_name in chunk.component_names() {
+            let chunk = Arc::new(chunk.range(&query, component_name));
+            let Some(num_events) = chunk.num_events_for_component(component_name) else {
+                continue;
+            };
+            let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                continue;
+            };
+            visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+        }
+    }
+}
+
+fn visit_relevant_chunks(
+    db: &re_entity_db::EntityDb,
+    entity_path: &EntityPath,
+    component_name: Option<ComponentName>,
+    timeline: Timeline,
+    time_range: ResolvedTimeRange,
+    mut visitor: impl FnMut(Arc<Chunk>, &ChunkTimeline, usize),
+) {
+    let query = RangeQuery::new(timeline, time_range);
+
+    if let Some(component_name) = component_name {
+        let chunks = db
+            .store()
+            .range_relevant_chunks(&query, entity_path, component_name);
+
+        for chunk in chunks {
+            let Some(num_events) = chunk.num_events_for_component(component_name) else {
+                continue;
+            };
+
+            let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                continue;
+            };
+
+            visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+        }
+    } else {
+        let mut seen = HashSet::new();
+        if let Some(subtree) = db.tree().subtree(entity_path) {
+            subtree.visit_children_recursively(&mut |entity_path, _| {
+                let Some(components) = db.store().all_components(&timeline, entity_path) else {
+                    return;
+                };
+
+                for component_name in components {
+                    let chunks =
+                        db.store()
+                            .range_relevant_chunks(&query, entity_path, component_name);
+
+                    for chunk in chunks {
+                        let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                            continue;
+                        };
+
+                        if seen.contains(&chunk.id()) {
+                            continue;
+                        }
+                        seen.insert(chunk.id());
+
+                        visitor(
+                            Arc::clone(&chunk),
+                            chunk_timeline,
+                            chunk.num_events_cumulative(),
+                        );
+                    }
+                }
```

Comment:
We should have an API for finding the relevant chunks for a given entity, which would save a lot of CPU here.

@teh-cmc do you agree?

```suggestion
                let chunks =
                    db.store()
                        .range_relevant_chunks_for_entity(&query, entity_path);

                for chunk in chunks {
                    let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
                        continue;
                    };

                    visitor(
                        Arc::clone(&chunk),
                        chunk_timeline,
                        chunk.num_events_cumulative(),
                    );
                }
```